### ISSUE TITLE
Error when trying to run the nvidia/cuda:11.0-base container

### ISSUE THREAD

**Issue by cetiberiojr** [NVIDIA/nvidia-docker#1380]
Hi Guys,

I followed the instructions found here:

[Install Instructions](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#docker)

So, after installing all the nvidia-docker2 (I'm using ubuntu server 20.04 LTS).

I ran the command:

`sudo docker run --rm --gpus all nvidia/cuda:11.0-base nvidia-smi`

And I got this error:

`docker: Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"process_linux.go:432: running prestart hook 0 caused \\\"error running hook: exit status 1, stdout: , stderr: nvidia-container-cli: detection error: nvml error: function not found\\\\n\\\"\"": unknown.`

Searching here on the issues, I found one guy fixed using those commands:

[Issue](https://github.com/NVIDIA/nvidia-docker/issues/1280)

```
sudo apt-get install -y libnvidia-container1=1.0.7-1
sudo apt-get install -y libnvidia-container-tools=1.0.7-1
sudo apt-get install -y nvidia-container-toolkit=1.0.5-1
sudo apt-get install -y nvidia-container-runtime=3.1.4-1
sudo apt-get install -y nvidia-docker2=2.2.2-1
```

As I'm using the same driver as him (340.108)

```
Sat Sep  5 22:18:06 2020
+------------------------------------------------------+
| NVIDIA-SMI 340.108    Driver Version: 340.108        |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce 9800 GT     Off  | 0000:01:00.0     N/A |                  N/A |
| 43%   61C   P12    N/A /  N/A |     76MiB /  1023MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Compute processes:                                               GPU Memory |
|  GPU       PID  Process name                                     Usage      |
|=============================================================================|
|    0            Not Supported                                               |
+-----------------------------------------------------------------------------+
```

I tried to do the same thing, but without success, now I have another error:

`docker: Error response from daemon: exec: "nvidia-container-runtime-hook": executable file not found in $PATH.`

Can someone help me here?

**Comment by klueska** [comment]
You can't run a CUDA 11 container on a driver version with version 340.108. It should give a better error message, but my guess is that this is your issue.

Try using a CUDA 9 container instead (e.g. `nvidia/cuda:9.0-base`).

**Comment by klueska** [comment]
I take it back, it looks like for that old of a driver you will, in fact, need the older `nvidia-docker` stack.

Try running allowing downgrades when you install the stack:
```
sudo apt-get install -y --allow-downgrades libnvidia-container1=1.0.7-1
sudo apt-get install -y --allow-downgrades libnvidia-container-tools=1.0.7-1
sudo apt-get install -y --allow-downgrades nvidia-container-toolkit=1.0.5-1
sudo apt-get install -y --allow-downgrades nvidia-container-runtime=3.1.4-1
sudo apt-get install -y --allow-downgrades nvidia-docker2=2.2.2-1
```

**Comment by keppo070** [comment]
I have the same issue. I followed your suggestion about the downgrades but I cant seem to get it to work. 

Can I provide more info on the subject

```
Sat Nov 21 11:50:53 2020
+------------------------------------------------------+
| NVIDIA-SMI 340.108    Driver Version: 340.108        |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce 8600 GT     Off  | 0000:01:00.0     N/A |                  N/A |
|  0%   56C    P0    N/A /  N/A |      3MiB /   255MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Compute processes:                                               GPU Memory |
|  GPU       PID  Process name                                     Usage      |
|=============================================================================|
|    0            Not Supported                                               |
+-----------------------------------------------------------------------------+
```

```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:349: starting container process caused "process_linux.go:449: container init caused \"process_linux.go:432: running prestart hook 0 caused \\\"error running hook: exit status 1, stdout: , stderr: nvidia-container-cli: requirement error: unsatisfied condition: cuda>=11.1, please update your driver to a newer version, or use an earlier cuda container\\\\n\\\"\"": unknown.
```

```

libnvidia-container-tools/bionic,now 1.0.7-1 amd64 [installed,upgradable to: 1.3.0-1]
libnvidia-container1/bionic,now 1.0.7-1 amd64 [installed,upgradable to: 1.3.0-1]
nvidia-340/bionic-updates,bionic,now 340.108-0ubuntu0.18.04.1 amd64 [installed]
nvidia-container-runtime/bionic,now 3.1.4-1 amd64 [installed,upgradable to: 3.4.0-1]
nvidia-container-toolkit/bionic,now 1.0.5-1 amd64 [installed,upgradable to: 1.3.0-1]
nvidia-docker2/bionic,now 2.2.2-1 all [installed,upgradable to: 2.5.0-1]
nvidia-opencl-icd-340/bionic-updates,bionic,now 340.108-0ubuntu0.18.04.1 amd64 [installed,automatic]
nvidia-settings/unknown,unknown,now 455.45.01-0ubuntu1 amd64 [installed,automatic]

```

**Comment by klueska** [comment]
@keppo070 your error is different.

The original one was:
```
nvidia-container-cli: detection error: nvml error: function not found
```

Yours is:
```
nvidia-container-cli: requirement error: unsatisfied condition: cuda>=11.1, please update your driver to a newer version, or use an earlier cuda container
```

The only way you will get this to work on a `340.108` driver is to use an older CUDA image (e.g. `nvidia/cuda:9.0-runtime`).

**Comment by manupatet** [comment]
Hey @cetiberiojr, were you able to fix this? 

Hi @klueska 

Thanks for the pointers above. Unfortunately, no matter what I throw at it, I keep getting `nvidia-container-cli: detection error: nvml error: function not found`. I tried downgrading as suggested [here](https://github.com/NVIDIA/nvidia-docker/issues/1280) (@frederico-klein 's solution), different CUDA images (11, 9, 6), different driver versions even (screwed up the system), but all in vein.

My config is:
- GPU(s): `GeForce 8800 GTS 320MB` and `GeForce 8800 GTS 640MB`
- Driver version: `340.108`
- OS: `Ubuntu 20.04 LTS`
- CUDA image: `nvidia/cuda:6.0-base`

On directly running `nvidia-smi`, it spits out this (FWIW)
```
+------------------------------------------------------+                       
| NVIDIA-SMI 340.108    Driver Version: 340.108        |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce 8800 GTS    Off  | 0000:04:00.0     N/A |                  N/A |
| 60%   62C    P0    N/A /  N/A |    154MiB /   639MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce 8800 GTS    Off  | 0000:05:00.0     N/A |                  N/A |
| 60%   69C    P0    N/A /  N/A |      2MiB /   319MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Compute processes:                                               GPU Memory |
|  GPU       PID  Process name                                     Usage      |
|=============================================================================|
|    0            Not Supported                                               |
|    1            Not Supported                                               |
+-----------------------------------------------------------------------------+
```

Maybe its because my GPUs are too old? Because I see that Frederico's GPU is GTX-580, a Fermi arch, supposedly the [oldest supported GPUs](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#support-hardware)? Am I right?

Any help is greatly appreciated.

**Comment by elezar** [comment]
Hi @manupatet looking at the error message, I would assume that one of the calls [here](https://github.com/NVIDIA/libnvidia-container/blob/master/src/nvc_info.c#L566) is associated with an NVML function that was added in a version later than the one associated with driver version `340.108`. (this was also the cause of the issue #1280 although this specific case has subsequently been addressed).

Also, could you run `apt list --installed | grep nvidia` to ensure that the downgraded versions are installed as expected.

**Comment by manupatet** [comment]
Hi @elezar,

tl;dr: It works ! Thanks a million!

So I ran the command you provided and it listed some `version 460` entries there:
```
libnvidia-common-460/focal-updates,focal-updates,focal-security,focal-security,now 460.39-0ubuntu0.20.04.1 all [installed,auto-removable]
libnvidia-container-tools/bionic,now 1.3.3-1 amd64 [installed]
libnvidia-container1/bionic,now 1.3.3-1 amd64 [installed,automatic]
libnvidia-extra-460/focal-updates,focal-security,now 460.39-0ubuntu0.20.04.1 amd64 [installed,auto-removable]
nvidia-340/focal-updates,now 340.108-0ubuntu5.20.04.1 amd64 [installed]
nvidia-container-runtime/bionic,now 3.1.4-1 amd64 [installed,upgradable to: 3.4.2-1]
nvidia-container-toolkit/bionic,now 1.4.2-1 amd64 [installed]
nvidia-docker2/bionic,now 2.2.2-1 all [installed,upgradable to: 2.5.0-1]
nvidia-opencl-icd-340/focal-updates,now 340.108-0ubuntu5.20.04.1 amd64 [installed,automatic]
nvidia-settings/focal-updates,now 460.39-0ubuntu0.20.04.1 amd64 [installed,automatic]
```

Not wanting leave anything to chance, I decided to reinstall OS. The stack looked a bit different this time:

```
libnvidia-container-tools/bionic,now 1.0.7-1 amd64 [installed,upgradable to: 1.3.3-1]
libnvidia-container1/bionic,now 1.0.7-1 amd64 [installed,upgradable to: 1.3.3-1]
nvidia-340/focal-updates,now 340.108-0ubuntu5.20.04.1 amd64 [installed]
nvidia-container-runtime/bionic,now 3.1.4-1 amd64 [installed,upgradable to: 3.4.2-1]
nvidia-container-toolkit/bionic,now 1.0.5-1 amd64 [installed,upgradable to: 1.4.2-1]
nvidia-docker2/bionic,now 2.2.2-1 all [installed,upgradable to: 2.5.0-1]
nvidia-opencl-icd-340/focal-updates,now 340.108-0ubuntu5.20.04.1 amd64 [installed,automatic]
nvidia-settings/focal-updates,now 460.39-0ubuntu0.20.04.1 amd64 [installed,automatic]
```
It still didn't work, so I followed the [steps](https://github.com/NVIDIA/nvidia-docker/issues/1280#issuecomment-671505779) @frederico-klein mentioned, and boom it worked !

```
Tue Apr 13 17:24:10 2021       
+------------------------------------------------------+                       
| NVIDIA-SMI 340.108    Driver Version: 340.108        |                       
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce 8800 GTS    Off  | 0000:04:00.0     N/A |                  N/A |
| 60%   54C    P0    N/A /  N/A |     32MiB /   639MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Compute processes:                                               GPU Memory |
|  GPU       PID  Process name                                     Usage      |
|=============================================================================|
|    0            Not Supported                                               |
+-----------------------------------------------------------------------------+
```

Thank you so much for the hint! So heartening to see my 14 yr old GPUs cranking away 🥳
